OpenRA.Mods.Cnc/Traits/World/WithResourceAnimation.cs

```diff
@@ -0,0 +1,105 @@
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
+	[Desc("Allows to play animations on resources.", "Attach this to the world actor.")]
+	public class WithResourceAnimationInfo : TraitInfo, Requires<IResourceLayerInfo>
+	{
+		[FieldLoader.Require]
+		[Desc("Resource types to animate.")]
+		public readonly HashSet<string> Types = null;
+
+		[Desc("The percentage of resource cells to play the animation on.", "Use two values to randomize between them.")]
+		public readonly int[] Ratio = { 1, 5 };
+
+		[Desc("Tick interval between two animation spawning.", "Use two values to randomize between them.")]
+		public readonly int[] Interval = { 200, 500 };
+
+		[FieldLoader.Require]
+		[Desc("Animation image.")]
+		public readonly string Image = null;
+
+		[SequenceReference(nameof(Image))]
+		[Desc("Randomly select one of these sequences to render.")]
+		public readonly string[] Sequences = new string[] { "idle" };
+
+		[PaletteReference]
+		[Desc("Animation palette.")]
+		public readonly string Palette = null;
+
+		public override object Create(ActorInitializer init) { return new WithResourceAnimation(init.Self, this); }
+	}
+
+	class WithResourceAnimation : IWorldLoaded, ITick
+	{
+		readonly WithResourceAnimationInfo info;
+		readonly World world;
+
+		WorldRenderer worldRenderer;
+		IResourceRenderer resourceRenderer;
+
+		int ticks;
+
+		public WithResourceAnimation(Actor self, WithResourceAnimationInfo info)
+		{
+			world = self.World;
+			this.info = info;
+
+			ticks = Common.Util.RandomInRange(world.LocalRandom, info.Interval);
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		{
+			worldRenderer = wr;
+
+			resourceRenderer = w.WorldActor.TraitsImplementing<IResourceRenderer>()
+				.Single(r => info.Types.Overlaps(r.ResourceTypes));
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (--ticks > 0)
+				return;
+
+			var cells = new HashSet<CPos>();
+			foreach (var uv in worldRenderer.Viewport.AllVisibleCells.CandidateMapCoords)
+			{
+				if (!world.Map.Contains(uv))
+					return;
+
+				var cell = uv.ToCPos(world.Map);
+				var type = resourceRenderer.GetRenderedResourceType(cell);
+				if (type != null && info.Types.Contains(type))
+					cells.Add(cell);
+			}
+
+			var ratio = Common.Util.RandomInRange(world.LocalRandom, info.Ratio);
+			var positions = cells.Shuffle(world.LocalRandom)
+				.Take(cells.Count * ratio / 100)
```

Comment:
This means there will be no effect with <20 to <100 cells in the viewport and the default ratio interval, which makes this really hard to spot at all. Maybe we need to increase the ratio?

Reply:
Or maybe round up so there is always at least 1?

Reply:
I just stared 3.5 minutes at a patch on Ritual Circle until the first animation was triggered and then one more minute until the second. Maybe we should change this to always play 1 to X animations every interval and not allow 0?

Reply:
I increased the ratio a bit and rounded upwards.